portfolio/src/main/webapp/script.js

```diff
@@ -497,14 +455,160 @@ function locationToArray(){
-function getWaypointFromGrid() {
-    fetch('/grids?requestRow=0&requestCol=0')
-    .then(response => console.log(response.json()))
-    .then((waypoint) => {
-        console.log('Row 0 Col 0, waypoint: ' + waypoint);
+/** Receives an array of routes, finds the safest one using a rating system */
+function pickSafestRoute(routesFromSafeNeighborGrids){
+  var counter = 0;
+  var routeIndex = 0;
+  var lessCrimesInRoute = 1000;
+
+  for (var r = 0; r < routesFromSafeNeighborGrids.routes.length; r++) {
+    var route = result.routes[r];
+
+    for (var j = 0; j < routesFromSafeNeighborGrids.legs[0].steps.length; j++) {
+      var routeArray = new google.maps.Polyline({
+        path: [
+            new google.maps.LatLng(routesFromSafeNeighborGrids.legs[0].steps[j].start_location.lat(),
+              routesFromSafeNeighborGrids.legs[0].steps[j].start_location.lng()),
+            new google.maps.LatLng(routesFromSafeNeighborGrids.legs[0].steps[j].end_location.lat(), 
+              routesFromSafeNeighborGrids.legs[0].steps[j].end_location.lng())
+        ]
+      });
+      
+      fetch('/markers').then(response => response.json()).then((markers) => {
+        markers.forEach((marker) => {
+          var myPosition = new google.maps.LatLng(marker.lat, marker.lng);
+          
+          if (google.maps.geometry.poly.isLocationOnEdge(myPosition, routeArray, 0.0064)) {
+            counter += rateCrime(marker.crimeType);
+            console.log("Rating" + counter);
+            console.log(marker.lat);
+            console.log(marker.lng);
+          }
+        });
+      });
+    }
+    
+    console.log("Rating" + counter);
+    
+    if (counter < lessCrimesInRoute) {
+      lessCrimesInRoute = counter;
+      routeIndex = r;
+    }
+    
+    counter = 0;
+  }
+
+  console.log("Chosen route index: " + routeIndex);
+
+  new google.maps.DirectionsRenderer({
+    map: map,
+    directions: result,
+    routeIndex: routeIndex
+  });    
+}
+
+function getWaypointForGrid(grid) {
+  fetch('/grids?requestRow=' + grid.row + '&requestCol=' + grid.col)
+  .then(response => response.json())
+  .then((waypoint) => {
+    let waypointForGrid = {
+      location: new google.maps.LatLng(waypoint.lat, waypoint.lng),
+      stopover: true
+    }
+    
+  return waypointForGrid;
+  })
+  .catch((error) => {
+    console.error(error);
+  }); 
+}
+
+/** Return the safest neighborings grid found [up, down, right, left].
+    Assumption: at least one grid will be safe. */
+function getSafeNeighboringGrids(grid){
+  let upperGridRow = -1;
+  let lowerGridRow = -1;
+  let rightGridCol = -1;
+  let leftGridCol = -1;
+  let safeGrids = [];
+  let safeGridUp;
+  let safeGridDown;
+  let safeGridRight;
+  let safeGridLeft;
+  let index = 0;
+
+  if(grid.row + 1 < 6){
+    upperGridRow = grid.row + 1;
+
+    // Might need to change the parameters once the class for the numberOfReports is merged.
+    fetch('/numberOfReports?row=' + upperGridRow + '&col=' + grid.col)
```

Comment:
same here. this will not block without await. you should write some dummy code to make sure it works

var grid = new Object();
grid.row = 1;
grid.col = 2;
console.log(getSafeNeightboringGrids(grid));

Reply:
Okay, I'll add it, thank you Yi-Tao!